aptos-move/framework/aptos-framework/sources/TokenCoinSwap.move

```diff
@@ -0,0 +1,231 @@
+// A basic utility module for swapping a fixed amount of coins with a fixed amount of tokens.
+module AptosFramework::TokenCoinSwap {
+    use Std::Signer;
+    use Std::Option::Option;
+    use AptosFramework::Table::{Self, Table};
+    use AptosFramework::Token::{Self, Token, TokenId};
+    use AptosFramework::Coin::{Self, Coin};
+    use AptosFramework::SimpleMap::{SimpleMap, Self};
+    use AptosFramework::Timestamp;
+
+    const ETOKEN_ALREADY_LISTED: u64 = 1;
+    const ETOKEN_LISTING_NOT_EXIST: u64 = 2;
+    const ETOKEN_NOT_IN_ESCROW: u64 = 3;
+    const ETOKEN_CANNOT_MOVE_OUT_OF_ESCROW_BEFORE_LOCKUP_TIME: u64 = 4;
+    const ETOKEN_MIN_PRICE_NOT_MATCH: u64 = 5;
+    const ETOKEN_AMOUNT_NOT_MATCH: u64 = 6;
+    const ENOT_ENOUGH_COIN: u64 = 7;
+
+    struct TokenCoinSwap<phantom CoinType> has store, drop {
+        token_amount: u64,
+        min_price_per_token: u64,
+    }
+
+    // stored at token owner's account
+    struct TokenListings<phantom CoinType> has key {
+        // key is the token id for swapping and value is the min price of target coin type.
+        listings: Table<TokenId, TokenCoinSwap<CoinType>>
+    }
+
+    // stored at coin owner's account
+    struct CoinListings<phantom CoinType> has key {
+        // key is the token id for swapping and value is the min price of target coin type.
+        listings: Table<TokenId, TokenCoinSwap<CoinType>>
+    }
+
+    // stored at owner's account
+    struct CoinEscrow<phantom CoinType> has key {
+        coin: Coin<CoinType>,
+        // until the locked time runs out, the owner cannot move the coin out of the escrow
+        // the default value is 0 meaning the owner can move the coin out anytime
+        locked_until_secs: u64,
+    }
+
+    struct TokenEscrow has store {
+        token: Token,
+        // until the locked time runs out, the owner cannot move the token out of the escrow
+        // the default value is 0 meaning the owner can move the coin out anytime
+        locked_until_secs: u64
+    }
+
+    // stored at owner's account
+    struct TokenStoreEscrow has key {
+        token_escrows: Table<TokenId, TokenEscrow>,
+    }
+
+    public fun exchange_coin_for_token<CoinType>(
+        coin_owner: &signer,
+        coin_amount: u64,
+        token_owner: address,
+        token_id: TokenId,
+        token_amount: u64
+    ) acquires TokenListings, TokenStoreEscrow {
+        // valide listing existing and coin owner has sufficient balance
+        let token_listing = borrow_global_mut<TokenListings<CoinType>>(token_owner);
+        assert!(Table::contains(&token_listing.listings, token_id), ETOKEN_LISTING_NOT_EXIST);
+        assert!(Coin::balance<CoinType>(Signer::address_of(coin_owner)) >= coin_amount, ENOT_ENOUGH_COIN);
+        // validate min price and amount
+        let token_swap = Table::borrow_mut(&mut token_listing.listings, token_id);
+        assert!(token_swap.min_price_per_token * token_amount <= coin_amount, ETOKEN_MIN_PRICE_NOT_MATCH);
```

Comment:
This can abort with multiplication overflow. We should decide whether we allow this in the fx. For diem, the goal was that every abort needs to be explicit via an assert. (The prover was verifying this for you.)

Reply:
Yes, it will abort with some error info like.  For this particular use, I think it is fine if the fx abort.
`VMError (if there is one): VMError {
│     major_status: ARITHMETIC_ERROR,
...

Reply:
SG, lets table this discussion until we have more clarity for the overall revised error model.